python/telemetry/snet.py

```diff
@@ -168,10 +100,91 @@ def _decode(self, obj, context, path=None):
     'TargetData_ControlIsActive' / Int8ul  # flag, really
     )
 
+LTUFrameHeader = BitStruct(
+    'SrcId' / BitsInteger(7),
+    'DstId' / BitsInteger(7),
+    'FrCntTx' / BitsInteger(4),
+    'FrCntRx' / BitsInteger(4),
+    'SNR' / BitsInteger(4),
+    'AiTypeSrc' / BitsInteger(4),
+    'AiTypeDst' / BitsInteger(4),
+    'DfcId' / BitsInteger(2),
+    'Caller' / Flag,
+    'Arq' / Flag,
+    'PduTypeId' / Flag,
+    'BchRq' / Flag,
+    'Hailing' / Flag,
+    'UdFl1' / Flag,
+    'PduLength' / BitsInteger(10),
+    'CRC13' / BitsInteger(13),
+    'CRC5' / BitsInteger(5),
+    Padding(2)
+    )
+
+# For reference:
+#
+# To send mailbox messages
+# FCIDMajor = 56  (COM0)
+# FCIDSub = 34 || 134  (COM1)
+
+MailboxTelemetry = Struct(
+    'SysTime' / BitsInteger(4),
+    'MemDataType' / BitsInteger(1),
+    'MemData32' / Hex(GreedyBytes)
+    # 'MemData32' / LinearAdapter(32, Int8ub)
+    )
+
 
-snet = Struct(
-    'header' / SNETFrameHeader,
-    'telemetry' / Switch(lambda c: (c.header.FCIDMajor, c.header.FCIDSub), {
+class TimeAdapter(Adapter):
+    def _encode(self, obj, context, path=None):
+        d = int((obj - datetime.datetime(2000, 1, 1))*2)
+        return Container(
+            days=d.days,
+            milliseconds=d.seconds * 1000 + d.microseconds / 1000)
+
+    def _decode(self, obj, context, path=None):
+        return (datetime.datetime(2000, 1, 1)
+                + datetime.timedelta(seconds=float(obj)/2.0))
+
+
+TimeStamp = TimeAdapter(BitsInteger(32, swapped=True))
+
+PDUFrameHeaderExtension = BitStruct(
+    'VersNo' / BitsInteger(2),
+    'DFCID' / BitsInteger(2),
+    'ExtensionRFU' / BitsInteger(4),
+    'ChannelInfo' / BitsInteger(8),
+    'QoS' / Flag,
+    'PDUTypeID' / Flag,
+    'ARQ' / Flag,
+    'ControlRFU' / BitsInteger(5),
+    'TimeTagSub' / BitsInteger(16),
+    'SCID' / BitsInteger(10),
+    'SeqNo' / BitsInteger(14)
+    )
+
+PDUFrameHeader = BitStruct(
+    Const(0b111100110101000000, BitsInteger(18)),
+    'CRC' / BitsInteger(14),
+    'FCIDMajor' / BitsInteger(6),
+    'FCIDSub' / BitsInteger(10),
+    'Urgent' / Flag,
+    'Extended' / Flag,
+    'CheckCRC' / Flag,
+    'Multiframe' / Flag,
+    'TimeTaggedSetting' / Flag,
+    'TimeTagged' / Flag,
+    'DataLength' / BitsInteger(10),
+    'TimeTag' / If(lambda c: c.TimeTagged, TimeStamp),
+    'Extension' / If(lambda c: c.Extended, PDUFrameHeaderExtension),
+    'UserData' / Switch(lambda c: (c.FCIDMajor, c.FCIDSub), {
         (0, 0): ADCSTelemetry,
         (9, 0): EPSTelemetry,
-    }, default=Pass))
+        (54, 17): MailboxTelemetry,  # COM0
+        (54, 117): MailboxTelemetry  # COM1
+    },
+      # why parse instead of PASS? it would be good to have a repesentation
+      # in case the wave file contains a new pair of FCIDMajor and FCIDSub not
+      # supported. It creates a bit of lag, but I guess is worth it instead of
+      # losing the UserData.
+      default=GreedyString("utf-8")))
```

Comment:
Would this cause problems if the binary data is not valid as a UTF-8 string?

Reply:
That is a good point, as I don't have access or know what the structure is for the rest of the telemetry frames, most likely will not be the best way to parse in this way. As I am not entirely familiar with the construct library and this scenarios, what do you think would be the safe/best type to use to catch the `UserData` sections when they are not a case for the `Switch()`?

The only thing I know as a general rule for `UserData` is its size: 256 Bytes.

Reply:
See https://github.com/daniestevez/gr-satellites/pull/508#discussion_r1274123176

Reply:
this was a bad take from my part, I switched it to catch the size of the `Payload` defined as 256 bytes